lib/binary/binary.nit

```diff
@@ -0,0 +1,303 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Add reading and writing binary services
+#
+# ~~~
+# var w = new FileWriter.open("/tmp/data.bin")
+# w.write "hello"
+# w.write_int64 123456789
+# w.write_byte 3
+# w.write_float 1.25
+# w.write_double 1.234567
+# w.write_bits(true, false, true)
+# assert w.last_error == null
+# w.close
+#
+# var r = new FileReader.open("/tmp/data.bin")
+# assert r.read(5) == "hello"
+# assert r.read_int64 == 123456789
+# assert r.read_byte == 3
+# assert r.read_float == 1.25
+# assert r.read_double == 1.234567
+#
+# var bits = r.read_bits
+# assert bits[0] and not bits[1] and bits[2]
+#
+# assert r.last_error == null
+# r.close
+# ~~~
+module binary
+
+in "C" `{
+	#include <inttypes.h>
+	#include <endian.h>
+
+	// Android compatibility
+	#ifndef be64toh
+		#define be64toh(val) betoh64(val)
+	#endif
+	#ifndef le64toh
+		#define le64toh(val) letoh64(val)
+	#endif
+`}
+
+# A stream of binary data
+abstract class BinaryStream
+	super Stream
+
+	# Write/read integer values as big-endian? `true` by default
+	var big_endian = true is writable
+end
+
+redef abstract class Writer
+	super BinaryStream
+
+	# Write a boolean `value` on a byte, using 0 for `false` and 1 for `true`
+	fun write_bool(value: Bool) do write_byte if value then 1 else 0
+
+	# Write up to 8 `Bool` in a byte
+	#
+	# To be used with `BinaryReader::read_bits`.
+	#
+	# Ensure: `bits.length <= 8`
+	fun write_bits(bits: Bool...)
+	do
+		assert bits.length <= 8
+
+		var int = 0
+		for b in bits.length.times do
+			if bits[b] then int += 2**b
+		end
+
+		write_byte int
+	end
+
+	# Write a floating point `value` on 32 bits
+	#
+	# Using this format may result in a loss of precision as it uses less bits
+	# than Nit `Float`.
+	fun write_float(value: Float)
+	do
+		for i in [0..4[ do write_byte value.float_byte_at(i)
+	end
+
+	# Write a floating point `value` on 64 bits
+	fun write_double(value: Float)
+	do
+		for i in [0..8[ do write_byte value.double_byte_at(i)
+	end
+
+	# Write `value` as a signed integer on 64 bits
+	#
+	# Using this format may result in a loss of precision as the length of a
+	# Nit `Int` may be more than 64 bits on some platforms.
+	fun write_int64(value: Int)
+	do
+		for i in [0..8[ do write_byte value.int64_byte_at(i, big_endian)
+	end
+
+	# TODO:
+	#
+	# fun write_int8
+	# fun write_uint8 # No need for this one, it is the current `read_char`
+	# fun write_int16
+	# fun write_uint16
+	# fun write_int32
+	# fun write_uint32
+	# fun write_uint64
+	# fun write_long_double?
+end
+
+redef abstract class Reader
+	super BinaryStream
+
+	# Read a single byte and return `true` if its value is different than 0
+	fun read_bool: Bool do return read_byte != 0
+
+	# Get an `Array` of 8 `Bool` by reading a single byte
+	#
+	# To be used with `BinaryWriter::write_bits`.
+	fun read_bits: Array[Bool]
```

Comment:
Hum, interesting... the `bitset` module add features to use an `Int` as a set of bits. That `bitset` can be used in combination with `write_int64`. `read_bits` and `write_bits` are a bit more friendly and over a byte instead of a long.